### PR TITLE
Download tentacles before tests start, to avoid tests timing out downloading tentacles.

### PR DESCRIPTION
**Pull request by LukeButters**
# Background

Some tests would timeout because the tests themselves download tentacles on demand. This can be slow taking longer than `1m` resulting in some tests not being able to start in time. This PR updates the test run such that we download tentacles first and then run the tests.

Other notes/changes
* Setup fixtures in nunit suck.
    * `[SetUpFixture]` in the wrong name space wont be called so the `TentacleIntegrationSetupFixtures` is in the required namespace.
    * If more than one `[SetUpFixture]` exists, we will only get logging from one `[SetUpFixture]`. Hence `TentacleIntegrationSetupFixtures` is the only `[SetUpFixture]` and will call other `ISetupFixture`s a type added in this PR.
* Logging is updated to locally log to `TestContext.Progress` unless in teamcity or turned off.
* Reading environment variables is cached to avoid needing to read an environment variable for each line logged.
* Tentacle fetching logs to a serilog logger
* `[SetUpFixture]` one time setup logs are recorded to a stringbuffer and logged out in a test so that in teamcity we get some logging (sometimes).
* Moved tentacle fetching classes into a single directory for sanity.

[SC-56842}

# Before

```
The test timed out.
   at Octopus.Tentacle.Tests.Integration.Support.IntegrationTest.<SetUp>b__10_0() in C:\BuildAgent\work\639265b01610d682\source\Octopus.Tentacle.Tests.Integration\Support\IntegrationTest.cs:line 28
1)    at Octopus.Tentacle.Tests.Integration.Support.IntegrationTest.<SetUp>b__10_0() in C:\BuildAgent\work\639265b01610d682\source\Octopus.Tentacle.Tests.Integration\Support\IntegrationTest.cs:line 28
------- Stdout: -------
00:00:00.0000342 [ClientFileTransfersAreRetriedWhenRetriesAreEnabled] [FailedDownloadsAreRetriedAndIsEventuallySuccessful(Polling,6.3.417,Sync)] Test started
00:00:00.0013965 [TentacleBinaryCache] [FailedDownloadsAreRetriedAndIsEventuallySuccessful(Polling,6.3.417,Sync)] Will download tentacle: "6.3.417"
Downloading https://f.feedz.io/octopus-deploy/dependencies/packages/Octopus.Tentacle.CrossPlatformBundle/6.3.417/download to C:\Users\agentuser\AppData\Local\Octopus\Temp\d4de9e10-1e90-4d81-8b7b-e7d1cbed7e25\cc2263398d2241028d4dee2425094ff7
Downloading https://f.feedz.io/octopus-deploy/dependencies/packages/Octopus.Tentacle.CrossPlatformBundle/6.3.417/download (469363303 bytes)
Downloading Completed 27.5013711500151%
Downloading Completed 47.3581395433464%
Downloading Completed 56.3780726589953%
Downloading Completed 85.9337126319822%
Download Finished
```

# After

image

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Octopus.Tentacle.Tests.Integration/BumpThreadPoolForAllTests.cs [hunk 1]
```diff
@@ -1,23 +0,0 @@
-﻿using System;
-using System.Threading;
-using NUnit.Framework;
-
-namespace Octopus.Tentacle.Tests.Integration
-{
-    public class BumpThreadPoolForAllTests
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Tests.Integration/BumpThreadPoolForAllTests.cs:7]: Not deleted it is moved

File: source/Octopus.Tentacle.Tests.Integration/Support/TentacleFetcher.cs [hunk 2]
```diff
@@ -1,14 +0,0 @@
-using System;
-using System.Threading;
-using System.Threading.Tasks;
-
-namespace Octopus.Tentacle.Tests.Integration.Support
-{
-    public static class TentacleFetcher
-    {
-        public static async Task<string> GetTentacleVersion(string downloadPath, Version version, CancellationToken cancellationToken)
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Tests.Integration/Support/TentacleFetcher.cs:9]: Moved

File: source/Octopus.Tentacle.Tests.Integration/Support/TentacleFetcherFactory.cs [hunk 3]
```diff
@@ -1,18 +0,0 @@
-using System;
-using Octopus.Tentacle.Util;
-
-namespace Octopus.Tentacle.Tests.Integration.Support
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Tests.Integration/Support/TentacleFetcherFactory.cs:4]: moved

**Review by acodrington — APPROVED** [review]
Nice work, a bunch of hoops to jump through, but this is a very nice improvement.

**Pull request merged**